identityfederation.go

```diff
@@ -0,0 +1,184 @@
+type TokenExchangeResponse struct {
+	AccessToken string `json:"access_token"`
+	TokenType   string `json:"token_type"`
+	ExpiresIn   int    `json:"expires_in"` // in seconds
+	Scope       string `json:"scope"`
+}
+
+// jwtClaims represents the claims in a JWT token (minimal set for validation).
+type jwtClaims struct {
+	Exp int64 `json:"exp"`
+}
+
+// tokenTransport implements http.RoundTripper and handles token exchange and caching.
+type tokenTransport struct {
+	config    IdentityFederationConfig
+	transport http.RoundTripper
+
+	idToken        string
+	apiAccessToken string
+	expiresAt      time.Time
+
+	tokenRefreshMu sync.Mutex
+}
+
+// HTTPClient constructs an HTTP client that authenticates using identity federation.
+// The client automatically handles token exchange and caching.
+func (c IdentityFederationConfig) HTTPClient() (*http.Client, error) {
+	if c.ClientID == "" {
+		return nil, fmt.Errorf("ClientID is required")
+	}
+	if c.IDTokenFunc == nil {
+		return nil, fmt.Errorf("IDTokenFunc is required")
+	}
+	if c.BaseURL == "" {
+		c.BaseURL = defaultBaseURL.String()
+	}
+
+	transport := &tokenTransport{
+		config:    c,
+		transport: http.DefaultTransport,
+	}
+
+	// Perform initial token exchange to validate configuration early
+	if err := transport.refreshToken(context.Background()); err != nil {
+		return nil, err
+	}
+
+	return &http.Client{
+		Transport: transport,
+		Timeout:   defaultHttpClientTimeout,
+	}, nil
+}
+
+// RoundTrip executes a single HTTP transaction, refreshing the API access token if necessary.
+func (t *tokenTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	if t.apiAccessToken == "" || time.Now().After(t.expiresAt) {
+		if err := t.refreshToken(req.Context()); err != nil {
+			return nil, err
+		}
+	}
+
+	reqClone := req.Clone(req.Context())
+	reqClone.Header.Set("Authorization", "Bearer "+t.apiAccessToken)
+
+	return t.transport.RoundTrip(reqClone)
+}
+
+// refreshToken performs the token exchange and updates the cached token.
+func (t *tokenTransport) refreshToken(ctx context.Context) error {
+	t.tokenRefreshMu.Lock()
+	defer t.tokenRefreshMu.Unlock()
+
+	if t.idToken == "" || validateIDToken(t.idToken) != nil {
+		idToken, err := t.config.IDTokenFunc()
+		if err != nil {
+			return fmt.Errorf("failed to fetch ID token: %w", err)
+		}
+		if err := validateIDToken(idToken); err != nil {
+			return fmt.Errorf("fetched ID token is invalid: %w", err)
+		}
+		t.idToken = idToken
+	}
+
+	exchangeURL := fmt.Sprintf("%s/api/v2/oauth/token-exchange", t.config.BaseURL)
+	values := url.Values{
+		"client_id": {t.config.ClientID},
+		"jwt":       {t.idToken},
+	}.Encode()
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, exchangeURL, strings.NewReader(values))
+	if err != nil {
+		return fmt.Errorf("failed to create token exchange request: %w", err)
+	}
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+	client := &http.Client{
```

Comment:
Consider moving this http.Client to the `tokenTransport` type so that you can reuse it (and therefore potentially reuse underlying TCP connections).